File: ios/App/Sources/Sheets.swift
import RadarKit
import SwiftUI

// MARK: - Search

struct SearchSheet: View {
    @ObservedObject var model: DriveModel
    @FocusState private var focused: Bool

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            HStack(spacing: 12) {
                Image(systemName: "magnifyingglass")
                    .foregroundStyle(.white.opacity(0.7))

                TextField(
                    "",
                    text: Binding(
                        get: { model.searchQuery },
                        set: { model.onSearchQueryChanged($0) }
                    ),
                    prompt: Text("Search an address").foregroundStyle(.white.opacity(0.45))
                )
                .font(.system(size: 19))
                .foregroundStyle(.white)
                .textInputAutocapitalization(.words)
                .autocorrectionDisabled()
                .submitLabel(.search)
                .focused($focused)

                if model.searching {
                    ProgressView().controlSize(.small)
                }
            }
            .padding(.horizontal, 18)
            .padding(.vertical, 14)
            .radarGlassCapsule(interactive: false)
            .padding(20)

            if model.searchResults.isEmpty, model.searchQuery.count >= 3, !model.searching {
                Text("Nothing found")
                    .font(.subheadline)
                    .foregroundStyle(.white.opacity(0.5))
                    .padding(.horizontal, 24)
            }

            List(model.searchResults) { place in
                Button {
                    model.pickDestination(place)
                } label: {
                    VStack(alignment: .leading, spacing: 3) {
                        Text(place.name)
                            .font(.system(size: 16, weight: .medium))
                            .foregroundStyle(.white)
                        if let address = place.address {
                            Text(address)
                                .font(.caption)
                                .foregroundStyle(.white.opacity(0.55))
                        }
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)
                    .contentShape(Rectangle())
                }
                .listRowBackground(Color.clear)
                .listRowSeparatorTint(.white.opacity(0.12))
            }
            // The list is content, not chrome, so it gets no glass of its own —
            // glass on a scrolling list is the fastest way to make it unreadable.
            .listStyle(.plain)
            .scrollContentBackground(.hidden)

            Spacer(minLength: 0)
        }
        .onAppear { focused = true }
    }
}

// MARK: - Route preview

struct RoutePreviewSheet: View {
    @ObservedObject var model: DriveModel

    private var fastest: Double {
        model.routeChoices.map(\.option.durationS).min() ?? 0
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(model.destination?.name ?? "Building a route…")
                .font(.title3.weight(.semibold))
                .foregroundStyle(.white)

            if let address = model.destination?.address {
                Text(address)
                    .font(.caption)
                    .foregroundStyle(.white.opacity(0.55))
                    .padding(.top, 4)
            }

            Spacer().frame(height: 16)

            if model.routeChoices.isEmpty {
                HStack(spacing: 12) {
                    ProgressView().controlSize(.small)
                    Text("Working out the way there")
                        .font(.subheadline)
                        .foregroundStyle(.white.opacity(0.7))
                }
            } else {
                GlassGroup(spacing: 10) {
                    VStack(spacing: 8) {
                        ForEach(model.routeChoices) { choice in
                            RouteChoiceRow(
                                choice: choice,
                                isSelected: choice.id == model.selectedRoute,
                                extraSeconds: choice.option.durationS - fastest
                            ) {
                                model.selectRoute(choice.id)
                            }
                        }
                    }
                }
            }

            Spacer(minLength: 16)

            GlassGroup(spacing: 12) {
                HStack(spacing: 12) {
                    GlassActionButton(title: "Cancel") { model.endNavigation() }
                    GlassActionButton(
                        title: "Start",
                        systemImage: "location.fill",
                        tint: RadarGlass.route,
                        enabled: !model.routeChoices.isEmpty
                    ) {
                        model.startNavigation()
                    }
                }
            }
        }
        .padding(24)
    }
}

/// One alternative.
///
/// The camera and hazard count is the reason this picker exists: comparing
/// routes on time alone throws away the one axis this app knows about, and
/// "three minutes longer, two fewer cameras" is a trade a driver can actually
/// make.
private struct RouteChoiceRow: View {
    let choice: RouteChoice
    let isSelected: Bool
    let extraSeconds: Double
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack {
                VStack(alignment: .leading, spacing: 2) {
                    HStack(alignment: .lastTextBaseline, spacing: 8) {
                        Text(RouteTracker.formatDuration(choice.option.durationS))
                            .font(.system(size: 20, weight: .bold, design: .rounded))
                            .monospacedDigit()
                        Text(RouteTracker.formatDistance(choice.option.distanceM))
                            .font(.caption)
                            .foregroundStyle(.white.opacity(0.6))
                        if extraSeconds >= 60 {
                            Text("+\(Int(extraSeconds / 60)) min")
                                .font(.caption)
                                .foregroundStyle(.white.opacity(0.5))
                        }
                    }
                    Text(choice.threatSummary ?? "Nothing on this one")
                        .font(.caption.weight(.medium))
                        .foregroundStyle(
                            choice.threatSummary == nil
                                ? RadarGlass.affirm : RadarGlass.caution
                        )
                    if let traffic = RouteTracker.describeTraffic(choice.option) {
                        Text(traffic)
                            .font(.caption2)
                            .foregroundStyle(Color(red: 0.95, green: 0.6, blue: 0.29))
                    }
                }
                Spacer()
                if isSelected {
                    Image(systemName: "checkmark")
                        .font(.headline)
                }
            }
            .foregroundStyle(.white)
            .padding(.horizontal, 16)
            .padding(.vertical, 12)
            .contentShape(RoundedRectangle(cornerRadius: 16, style: .continuous))
        }
        .buttonStyle(.plain)
        .radarGlassPanel(cornerRadius: 16, tint: isSelected ? RadarGlass.route : nil)
    }
}

// MARK: - Reporting

/// Deliberately six large buttons and nothing else. Anything that needs reading
/// does not belong on a screen used at 100 km/h.
struct ReportSheet: View {
    let onPick: (String) -> Void

    private struct Option: Identifiable {
        let id: String
        let label: String
        let icon: String
    }

    private let options = [
        Option(id: "police", label: "Police", icon: "shield.fill"),
        Option(id: "mobile_camera", label: "Mobile camera", icon: "camera.fill"),
        Option(id: "crash", label: "Crash", icon: "car.fill"),
        Option(id: "hazard", label: "Hazard", icon: "exclamationmark.triangle.fill"),
        Option(id: "object_on_road", label: "Object on road", icon: "shippingbox.fill"),
        Option(id: "stopped_vehicle", label: "Stopped vehicle", icon: "car.side.fill"),
    ]

    var body: some View {
        VStack(alignment: .leading, spacing: 18) {
            Text("What did you just pass?")
                .font(.title3.weight(.semibold))
                .foregroundStyle(.white)

            GlassGroup(spacing: 12) {
                LazyVGrid(
                    columns: Array(repeating: GridItem(.flexible(), spacing: 10), count: 3),
                    spacing: 10
                ) {
                    ForEach(options) { option in
                        Button { onPick(option.id) } label: {
                            VStack(spacing: 8) {
                                Image(systemName: option.icon).font(.title2)
                                Text(option.label)
                                    .font(.caption)
                                    .multilineTextAlignment(.center)
                            }
                            .frame(maxWidth: .infinity)
                            .frame(height: 88)
                            .contentShape(RoundedRectangle(cornerRadius: 18, style: .continuous))
                        }
                        .buttonStyle(.plain)
                        .foregroundStyle(.white)
                        .radarGlassPanel(cornerRadius: 18)
                    }
                }
            }

            Spacer(minLength: 0)
        }
        .padding(24)
    }
}

// MARK: - Threat detail

struct ThreatSheet: View {
    let threat: Threat
    @ObservedObject var model: DriveModel

    /// Only community reports can be voted on. A government feed is not
    /// something a driver confirms or denies.
    private var isCommunityReport: Bool { threat.id.hasPrefix("community:") }

    private var provenance: String {
        if threat.isCamera { return "From published camera data" }
        if isCommunityReport { return "Reported by a driver" }
        return "From a live traffic feed"
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(AlertEngine.label(threat))
                .font(.title3.weight(.semibold))
                .foregroundStyle(.white)

            Text(provenance)
                .font(.subheadline)
                .foregroundStyle(.white.opacity(0.6))
                .padding(.top, 8)

            if let limit = AlertEngine.postedLimit(threat) {
                Text("Posted limit \(limit) km/h")
                    .font(.subheadline)
                    .foregroundStyle(.white.opacity(0.6))
                    .padding(.top, 6)
            }

            if isCommunityReport {
                Text("Is it still there?")
                    .font(.system(size: 15))
                    .foregroundStyle(.white)
                    .padding(.top, 22)

                GlassGroup(spacing: 12) {
                    HStack(spacing: 12) {
                        GlassActionButton(
                            title: "Still there",
                            systemImage: "checkmark",
                            tint: RadarGlass.affirm
                        ) {
                            Task { await model.vote(on: threat, confirm: true) }
                        }
                        GlassActionButton(
                            title: "Gone",
                            systemImage: "xmark",
                            tint: RadarGlass.warning
                        ) {
                            Task { await model.vote(on: threat, confirm: false) }
                        }
                    }
                }
                .padding(.top, 12)
            }

            Spacer(minLength: 0)
        }
        .padding(24)
    }
}

File: android/app/src/main/kotlin/au/radar/app/ui/Sheets.kt
package au.radar.app.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Search
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.focus.FocusRequester
import androidx.compose.ui.focus.focusRequester
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import au.radar.core.AlertEngine
import au.radar.core.PlaceResult
import au.radar.core.RouteOption
import au.radar.core.RouteTracker
import au.radar.core.Threat

/**
 * Every sheet in the app sits on this: a dimming scrim that dismisses on tap,
 * and a glass panel anchored to the bottom within thumb reach.
 *
 * The scrim is a flat colour rather than a blur, for the reason in Glass.kt —
 * the map underneath is a SurfaceView and cannot be sampled by the Compose
 * layer at any price worth paying.
 */
@Composable
private fun GlassSheet(
    onDismiss: () -> Unit,
    content: @Composable () -> Unit,
) {
    Box(
        Modifier
            .fillMaxSize()
            .background(Color.Black.copy(alpha = 0.55f))
            .clickable(onClick = onDismiss),
        contentAlignment = Alignment.BottomCenter,
    ) {
        // Swallow taps on the panel itself so they do not reach the scrim.
        Box(Modifier.clickable(enabled = false, onClick = {})) {
            GlassPanel(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(12.dp),
                shape = RoundedCornerShape(28.dp),
                strong = true,
            ) {
                Column(Modifier.padding(20.dp)) { content() }
            }
        }
    }
}

@Composable
private fun SheetTitle(text: String) {
    Text(text, fontSize = 19.sp, fontWeight = FontWeight.SemiBold, color = Color.White)
}

// MARK: - Search

@Composable
fun SearchSheet(
    query: String,
    results: List<PlaceResult>,
    searching: Boolean,
    onQueryChange: (String) -> Unit,
    onPick: (PlaceResult) -> Unit,
    onDismiss: () -> Unit,
) {
    val focus = remember { FocusRequester() }
    LaunchedEffect(Unit) { focus.requestFocus() }

    GlassSheet(onDismiss = onDismiss) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Icon(Icons.Filled.Search, contentDescription = null, tint = Color.White.copy(0.7f))
            Spacer(Modifier.width(12.dp))
            BasicTextField(
                value = query,
                onValueChange = onQueryChange,
                singleLine = true,
                textStyle = TextStyle(color = Color.White, fontSize = 19.sp),
                cursorBrush = SolidColor(Color(0xFF4AA8FF)),
                modifier = Modifier
                    .weight(1f)
                    .focusRequester(focus),
                decorationBox = { inner ->
                    if (query.isEmpty()) {
                        Text("Search an address", color = Color.White.copy(0.45f), fontSize = 19.sp)
                    }
                    inner()
                },
            )
            if (searching) {
                CircularProgressIndicator(
                    Modifier.size(20.dp),
                    strokeWidth = 2.dp,
                    color = Color.White.copy(0.7f),
                )
            }
        }

        Spacer(Modifier.height(16.dp))

        if (results.isEmpty() && query.length >= 3 && !searching) {
            Text("Nothing found", color = Color.White.copy(0.5f), fontSize = 15.sp)
        }

        LazyColumn(Modifier.heightIn(max = 340.dp)) {
            items(results) { place ->
                Column(
                    Modifier
                        .fillMaxWidth()
                        .clickable { onPick(place) }
                        .padding(vertical = 12.dp),
                ) {
                    Text(place.name, color = Color.White, fontSize = 16.sp, fontWeight = FontWeight.Medium)
                    place.address?.let {
                        Text(it, color = Color.White.copy(0.55f), fontSize = 13.sp)
                    }
                }
            }
        }
    }
}

// MARK: - Route preview

@Composable
fun RoutePreviewSheet(
    destination: PlaceResult?,
    route: RouteOption?,
    onStart: () -> Unit,
    onCancel: () -> Unit,
) {
    GlassSheet(onDismiss = onCancel) {
        SheetTitle(destination?.name ?: "Building a route…")
        destination?.address?.let {
            Spacer(Modifier.height(4.dp))
            Text(it, color = Color.White.copy(0.55f), fontSize = 13.sp)
        }

        Spacer(Modifier.height(18.dp))

        if (route == null) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                CircularProgressIndicator(Modifier.size(18.dp), strokeWidth = 2.dp, color = Color.White)
                Spacer(Modifier.width(12.dp))
                Text("Working out the way there", color = Color.White.copy(0.7f), fontSize = 15.sp)
            }
        } else {
            Row(verticalAlignment = Alignment.Bottom) {
                Text(
                    RouteTracker.formatDuration(route.durationS),
                    fontSize = 34.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color.White,
                )
                Spacer(Modifier.width(12.dp))
                Text(
                    RouteTracker.formatDistance(route.distanceM),
                    fontSize = 17.sp,
                    color = Color.White.copy(0.65f),
                    modifier = Modifier.padding(bottom = 5.dp),
                )
            }
        }

        Spacer(Modifier.height(20.dp))

        Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
            GlassActionButton(
                label = "Cancel",
                onClick = onCancel,
                modifier = Modifier.weight(1f),
            )
            GlassActionButton(
                label = "Start",
                onClick = onStart,
                modifier = Modifier.weight(1f),
                tint = Color(0xFF2F80ED),
                enabled = route != null,
            )
        }
    }
}

// MARK: - Reporting

/**
 * Deliberately six large buttons and nothing else. Anything that needs reading
 * does not belong on a screen used at 100 km/h.
 */
@Composable
fun ReportSheet(onPick: (String) -> Unit, onDismiss: () -> Unit) {
    data class Option(val kind: String, val label: String)

    val options = listOf(
        Option("police", "Police"),
        Option("mobile_camera", "Mobile camera"),
        Option("crash", "Crash"),
        Option("hazard", "Hazard"),
        Option("object_on_road", "Object on road"),
        Option("stopped_vehicle", "Stopped vehicle"),
    )

    GlassSheet(onDismiss = onDismiss) {
        SheetTitle("What did you just pass?")
        Spacer(Modifier.height(16.dp))

        LazyVerticalGrid(
            columns = GridCells.Fixed(3),
            horizontalArrangement = Arrangement.spacedBy(10.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
            modifier = Modifier.heightIn(max = 260.dp),
        ) {
            items(options) { option ->
                GlassPanel(
                    modifier = Modifier
                        .height(88.dp)
                        .fillMaxWidth()
                        .clickable { onPick(option.kind) },
                    shape = RoundedCornerShape(18.dp),
                ) {
                    Text(
                        option.label,
                        Modifier.align(Alignment.Center).padding(8.dp),
                        color = Color.White,
                        fontSize = 14.sp,
                        fontWeight = FontWeight.Medium,
                    )
                }
            }
        }
    }
}

// MARK: - Threat detail

@Composable
fun ThreatSheet(
    threat: Threat,
    onConfirm: () -> Unit,
    onDeny: () -> Unit,
    onDismiss: () -> Unit,
) {
    // Only community reports can be voted on. A government feed is not
    // something a driver confirms or denies.
    val isCommunityReport = threat.id.startsWith("community:")

    GlassSheet(onDismiss = onDismiss) {
        SheetTitle(AlertEngine.label(threat))

        Spacer(Modifier.height(8.dp))
        Text(
            when {
                threat.isCamera -> "From published camera data"
                isCommunityReport -> "Reported by a driver"
                else -> "From a live traffic feed"
            },
            color = Color.White.copy(0.6f),
            fontSize = 14.sp,
        )

        AlertEngine.postedLimit(threat)?.let { limit ->
            Spacer(Modifier.height(6.dp))
            Text("Posted limit $limit km/h", color = Color.White.copy(0.6f), fontSize = 14.sp)
        }

        if (isCommunityReport) {
            Spacer(Modifier.height(20.dp))
            Text("Is it still there?", color = Color.White, fontSize = 15.sp)
            Spacer(Modifier.height(12.dp))

            Row(horizontalArrangement = Arrangement.spacedBy(12.dp)) {
                GlassActionButton(
                    label = "Still there",
                    onClick = onConfirm,
                    modifier = Modifier.weight(1f),
                    tint = Color(0xFF27AE60),
                    icon = { Icon(Icons.Filled.Check, contentDescription = null) },
                )
                GlassActionButton(
                    label = "Gone",
                    onClick = onDeny,
                    modifier = Modifier.weight(1f),
                    tint = Color(0xFFEB5757),
                    icon = { Icon(Icons.Filled.Close, contentDescription = null) },
                )
            }
        }
    }
}

// MARK: - Shared

@Composable
private fun GlassActionButton(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    tint: Color = Color.Unspecified,
    enabled: Boolean = true,
    icon: (@Composable () -> Unit)? = null,
) {
    GlassPanel(
        modifier = modifier
            .height(54.dp)
            .clickable(enabled = enabled, onClick = onClick),
        shape = Glass.CapsuleShape,
        tint = tint,
        strong = true,
    ) {
        Row(
            Modifier.align(Alignment.Center),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (icon != null) {
                icon()
                Spacer(Modifier.width(8.dp))
            }
            Text(
                label,
                color = if (enabled) Color.White else Color.White.copy(0.4f),
                fontSize = 16.sp,
                fontWeight = FontWeight.SemiBold,
            )
        }
    }
}
